cpp/src/arrow/dataset/expression_benchmark.cc

```diff
@@ -0,0 +1,105 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <iostream>
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/cast.h"
+#include "arrow/dataset/expression.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/type.h"
+#include "arrow/util/range.h"
+
+namespace arrow {
+namespace dataset {
+
+static std::shared_ptr<Partitioning> MakePartitioning() {
+  auto schema = arrow::schema({
+      arrow::field("a", arrow::dictionary(arrow::int32(), arrow::int64())),
+      arrow::field("b", arrow::dictionary(arrow::int32(), arrow::int64())),
```

Comment:
nit: we don't need to be explicit about the arrow namespace here
```suggestion
  auto schema = schema({
      field("a", dictionary(int32(), int64())),
      feld("b", dictionary(int32(), int64())),
```